smu.c

```diff
@@ -557,22 +554,30 @@ dotable(const char *begin, const char *end, int newblock) {
 
 int
 doparagraph(const char *begin, const char *end, int newblock) {
-	const char *p;
-	regmatch_t match;
+	const char *p = NULL, *q = begin;
 
 	if (!newblock)
 		return 0;
-	if (regexec(&p_end_regex, begin + 1, 1, &match, 0)) {
-		p = end;
-	} else {
-		p = begin + 1 + match.rm_so;
+	while (!p) {
+		q = memchr(q, '\n', end - q);
```

Comment:
Shouldn't this be outside of the loop? If we don't find a newline between q and end, we don't have to do it again every time we step one character further.

Reply:
If we don't find a newline, `p` is set to `end` and thus the loop breaks. This is inside the loop in case we *do* find a newline, but it doesn't match "\n\n" or "\n```". Then we need to increment `q` and try again.